File: src/types/skill.ts
import { z } from "zod";

const SkillSourceTypeSchema = z.literal(["local_folder", "platform", "git"]);
const SkillSchema = z.object({
	/** Stable Skill Library identifier. */
	id: z.string().min(1),
	/** Folder name copied beneath project Skill snapshots. */
	folderName: z.string().min(1),
	/** User-visible Skill name. */
	displayName: z.string().min(1),
	/** Capability summary parsed from SKILL.md. */
	description: z.string(),
	/** Origin retained after the Skill is copied into managed storage. */
	sourceType: SkillSourceTypeSchema,
	/** Original source folder when one is available. */
	sourcePath: z.string().nullable(),
	/** Creation time in Unix milliseconds. */
	createdAtMs: z.int().nonnegative(),
	/** Latest metadata update time in Unix milliseconds. */
	updatedAtMs: z.int().nonnegative(),
});

const CompiledSkillSchema = z.compile(SkillSchema);
const CompiledSkillsSchema = z.compile(z.array(SkillSchema));

const EditorSkillFilesSchema = z.object({
	/** Editable UTF-8 files, keyed by relative path. */
	files: z.record(z.string(), z.string()),
	/** Complete directory list, including empty folders. */
	directories: z.array(z.string()),
	/** Destination-to-original paths for files that stay on disk. */
	retainedFiles: z.record(z.string(), z.string()),
});
const CompiledEditorSkillFilesSchema = z.compile(EditorSkillFilesSchema);
type EditorSkillFiles = z.infer<typeof EditorSkillFilesSchema>;

type Skill = z.infer<typeof SkillSchema>;

export type { EditorSkillFiles, Skill };
export {
	CompiledEditorSkillFilesSchema,
	CompiledSkillSchema,
	CompiledSkillsSchema,
};

File: src-tauri/src/domain/skill.rs
use std::path::PathBuf;

/// Origin metadata retained for a managed Skill.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(crate) enum SkillSourceType {
    LocalFolder,
    Platform,
    Git,
}

impl SkillSourceType {
    /// Returns the stable persisted and IPC identifier.
    pub(crate) const fn as_str(self) -> &'static str {
        match self {
            Self::LocalFolder => "local_folder",
            Self::Platform => "platform",
            Self::Git => "git",
        }
    }

    /// Parses an origin identifier read from trusted migrated storage.
    pub(crate) fn parse(value: &str) -> Option<Self> {
        match value {
            "local_folder" => Some(Self::LocalFolder),
            "platform" => Some(Self::Platform),
            "git" => Some(Self::Git),
            _ => None,
        }
    }
}

/// Theoria-managed Skill Library entry.
#[derive(Debug, Clone, PartialEq, Eq)]
pub(crate) struct Skill {
    /// Stable local identifier.
    pub(crate) id: String,
    /// Directory name copied into Task project Skill snapshots.
    pub(crate) folder_name: String,
    /// User-visible Skill name.
    pub(crate) display_name: String,
    /// Short capability description.
    pub(crate) description: String,
    /// How the Skill entered the Library.
    pub(crate) source_type: SkillSourceType,
    /// Managed path relative to application data.
    pub(crate) storage_relative_path: PathBuf,
    /// Original source directory when imported.
    pub(crate) source_path: Option<PathBuf>,
    /// Creation time in Unix milliseconds.
    pub(crate) created_at_ms: i64,
    /// Latest metadata update time in Unix milliseconds.
    pub(crate) updated_at_ms: i64,
}

/// Values inserted for a newly managed Skill.
#[derive(Debug, Clone, PartialEq, Eq)]
pub(crate) struct NewSkill {
    /// Stable identifier allocated before filesystem copying.
    pub(crate) id: String,
    /// Valid project Skill directory name.
    pub(crate) folder_name: String,
    /// User-visible Skill name.
    pub(crate) display_name: String,
    /// Short capability description.
    pub(crate) description: String,
    /// Import origin.
    pub(crate) source_type: SkillSourceType,
    /// Managed path relative to application data.
    pub(crate) storage_relative_path: PathBuf,
    /// Original import path.
    pub(crate) source_path: Option<PathBuf>,
    /// Creation and initial update time.
    pub(crate) created_at_ms: i64,
}
